File: sdk/python/objectio/_sigv4.py
"""AWS SigV4 for the S3 service, on the standard library alone.

Deliberately not botocore: this package is meant to be installable into an
operator or a provisioning job where pulling in the AWS SDK for one signing
function is a large dependency for no gain. The algorithm is stable.
"""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone

ALGORITHM = "AWS4-HMAC-SHA256"
TERMINATOR = "aws4_request"
SERVICE = "s3"

# Every request signs its payload hash; S3 wants the header whether or not
# there is a body, and an empty body hashes to this.
EMPTY_PAYLOAD_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"

_UNRESERVED = frozenset(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.~"
)


def escape(value: str) -> str:
    """Percent-encode everything outside the unreserved set.

    ``urllib.parse.quote`` is close but its default safe set and its handling
    of ``~`` have bitten people; spelling it out is shorter than the caveats.
    """
    out = []
    for byte in value.encode("utf-8"):
        ch = chr(byte)
        if ch in _UNRESERVED:
            out.append(ch)
        else:
            out.append(f"%{byte:02X}")
    return "".join(out)


def escape_path(path: str) -> str:
    """Encode a path the way SigV4 canonicalises it for S3: each segment
    escaped, the separators left alone, no double encoding.

    The same function builds the URL that is sent and the canonical path that
    is signed. That is the point — a signature mismatch on a key containing a
    space or a ``#`` is almost always those two disagreeing.
    """
    return "/".join(escape(seg) for seg in path.split("/"))


def canonical_query(params: dict[str, str] | None) -> str:
    if not params:
        return ""
    return "&".join(
        f"{escape(k)}={escape(v)}" for k, v in sorted(params.items())
    )


def _hmac(key: bytes, data: str) -> bytes:
    return hmac.new(key, data.encode("utf-8"), hashlib.sha256).digest()


def sign_headers(
    *,
    method: str,
    host: str,
    path: str,
    query: dict[str, str] | None,
    body: bytes,
    access_key: str,
    secret_key: str,
    region: str,
    content_type: str | None = None,
    now: datetime | None = None,
) -> dict[str, str]:
    """Return the headers that authenticate this request.

    ``path`` must be the *unescaped* path; it is canonicalised here so the
    caller cannot accidentally sign one spelling and send another.
    """
    stamp = (now or datetime.now(timezone.utc)).astimezone(timezone.utc)
    amz_date = stamp.strftime("%Y%m%dT%H%M%SZ")
    date_stamp = stamp.strftime("%Y%m%d")

    payload_hash = hashlib.sha256(body).hexdigest() if body else EMPTY_PAYLOAD_SHA256

    headers = {
        "host": host,
        "x-amz-content-sha256": payload_hash,
        "x-amz-date": amz_date,
    }
    if content_type:
        headers["content-type"] = content_type

    signed_names = sorted(headers)
    canonical_headers = "".join(f"{h}:{headers[h].strip()}\n" for h in signed_names)
    signed_headers = ";".join(signed_names)

    canonical_request = "\n".join(
        [
            method,
            escape_path(path) or "/",
            canonical_query(query),
            canonical_headers,
            signed_headers,
            payload_hash,
        ]
    )

    scope = f"{date_stamp}/{region}/{SERVICE}/{TERMINATOR}"
    string_to_sign = "\n".join(
        [
            ALGORITHM,
            amz_date,
            scope,
            hashlib.sha256(canonical_request.encode("utf-8")).hexdigest(),
        ]
    )

    key = _hmac(f"AWS4{secret_key}".encode("utf-8"), date_stamp)
    key = _hmac(key, region)
    key = _hmac(key, SERVICE)
    key = _hmac(key, TERMINATOR)
    signature = hmac.new(
        key, string_to_sign.encode("utf-8"), hashlib.sha256
    ).hexdigest()

    out = {
        "X-Amz-Date": amz_date,
        "X-Amz-Content-Sha256": payload_hash,
        "Authorization": (
            f"{ALGORITHM} Credential={access_key}/{scope}, "
            f"SignedHeaders={signed_headers}, Signature={signature}"
        ),
    }
    if content_type:
        out["Content-Type"] = content_type
    return out

File: sdk/go/objectio/sigv4.go
package objectio

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

// AWS SigV4 for the S3 service, implemented against the standard library.
//
// Deliberately not aws-sdk-go-v2: this SDK is meant to sit inside a CSI
// driver or an operator, where pulling the AWS SDK in for one signing
// function is a large dependency for no gain. The algorithm is stable and
// ~80 lines.

const (
	algorithm  = "AWS4-HMAC-SHA256"
	terminator = "aws4_request"
	service    = "s3"
	// Every request signs its payload hash. S3 requires the header whether or
	// not there is a body; an empty body hashes to this constant.
	emptyPayloadHash = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
)

// escapePath percent-encodes a path the way SigV4 canonicalises it for S3:
// each segment escaped, the separators left alone, and — unlike the generic
// signer — no double encoding.
//
// The same function produces the URL that is sent and the canonical path that
// is signed. That is the whole point: a signature mismatch on an object key
// containing a space or a '#' is almost always the two disagreeing.
func escapePath(path string) string {
	segments := strings.Split(path, "/")
	for i, s := range segments {
		segments[i] = escapeRFC3986(s)
	}
	return strings.Join(segments, "/")
}

// escapeRFC3986 escapes everything outside the unreserved set. url.QueryEscape
// is not usable here: it encodes a space as '+', which SigV4 requires to be
// %20.
func escapeRFC3986(s string) string {
	const unreserved = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.~"
	var b strings.Builder
	for i := 0; i < len(s); i++ {
		c := s[i]
		if strings.IndexByte(unreserved, c) >= 0 {
			b.WriteByte(c)
			continue
		}
		b.WriteByte('%')
		b.WriteString(strings.ToUpper(hex.EncodeToString([]byte{c})))
	}
	return b.String()
}

func canonicalQuery(q url.Values) string {
	keys := make([]string, 0, len(q))
	for k := range q {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(q))
	for _, k := range keys {
		values := append([]string(nil), q[k]...)
		sort.Strings(values)
		for _, v := range values {
			parts = append(parts, escapeRFC3986(k)+"="+escapeRFC3986(v))
		}
	}
	return strings.Join(parts, "&")
}

func hmacSHA256(key []byte, data string) []byte {
	h := hmac.New(sha256.New, key)
	h.Write([]byte(data))
	return h.Sum(nil)
}

func sha256Hex(b []byte) string {
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}

// sign adds the Authorization, X-Amz-Date and X-Amz-Content-Sha256 headers.
// `body` is the exact bytes that will be sent; pass nil for no body.
func sign(req *http.Request, accessKey, secretKey, region string, body []byte, now time.Time) {
	amzDate := now.UTC().Format("20060102T150405Z")
	dateStamp := now.UTC().Format("20060102")

	payloadHash := emptyPayloadHash
	if len(body) > 0 {
		payloadHash = sha256Hex(body)
	}

	host := req.Host
	if host == "" {
		host = req.URL.Host
	}
	req.Header.Set("X-Amz-Date", amzDate)
	req.Header.Set("X-Amz-Content-Sha256", payloadHash)

	// Sign host, the two x-amz headers, and content-type when one is set.
	// Anything signed must also be sent, and anything sent that is signed must
	// not be rewritten in flight — which is why content-length is left out.
	signed := []string{"host", "x-amz-content-sha256", "x-amz-date"}
	headers := map[string]string{
		"host":                 host,
		"x-amz-content-sha256": payloadHash,
		"x-amz-date":           amzDate,
	}
	if ct := req.Header.Get("Content-Type"); ct != "" {
		signed = append(signed, "content-type")
		headers["content-type"] = ct
	}
	sort.Strings(signed)

	var canonicalHeaders strings.Builder
	for _, h := range signed {
		canonicalHeaders.WriteString(h)
		canonicalHeaders.WriteByte(':')
		canonicalHeaders.WriteString(strings.TrimSpace(headers[h]))
		canonicalHeaders.WriteByte('\n')
	}
	signedHeaders := strings.Join(signed, ";")

	path := req.URL.EscapedPath()
	if path == "" {
		path = "/"
	}
	canonicalRequest := strings.Join([]string{
		req.Method,
		path,
		canonicalQuery(req.URL.Query()),
		canonicalHeaders.String(),
		signedHeaders,
		payloadHash,
	}, "\n")

	credentialScope := strings.Join([]string{dateStamp, region, service, terminator}, "/")
	stringToSign := strings.Join([]string{
		algorithm,
		amzDate,
		credentialScope,
		sha256Hex([]byte(canonicalRequest)),
	}, "\n")

	key := hmacSHA256([]byte("AWS4"+secretKey), dateStamp)
	key = hmacSHA256(key, region)
	key = hmacSHA256(key, service)
	key = hmacSHA256(key, terminator)
	signature := hex.EncodeToString(hmacSHA256(key, stringToSign))

	req.Header.Set("Authorization", algorithm+
		" Credential="+accessKey+"/"+credentialScope+
		", SignedHeaders="+signedHeaders+
		", Signature="+signature)
}
